admin/test/test_release.py

```diff
@@ -661,3 +667,637 @@ def test_publish_non_release_fails(self):
+        :param FakeYum yum: Fake yum utilities to interact with.
+        :param rpm_directory: See :py:func:`update_repo`.
+        :param target_bucket: See :py:func:`update_repo`.
+        :param target_key: See :py:func:`update_repo`.
+        :param source_repo: See :py:func:`update_repo`.
+        :param packages: See :py:func:`update_repo`.
+        :param version: See :py:func:`update_repo`.
+        """
+        dispatchers = [aws.get_dispatcher(), yum.get_dispatcher(),
+                       base_dispatcher]
+        sync_perform(
+            ComposedDispatcher(dispatchers),
+            upload_rpms(
+                scratch_directory=scratch_directory,
+                target_bucket=target_bucket,
+                version=version,
+                build_server=build_server,
+            ),
+        )
+
+    def update_repo(self, aws, yum,
+                    rpm_directory, target_bucket, target_key, source_repo,
+                    packages):
+        """
+        Call :func:``update_repo``, interacting with a fake AWS and yum
+        utilities.
+
+        :param FakeAWS aws: Fake AWS to interact with.
+        :param FakeYum yum: Fake yum utilities to interact with.
+        :param rpm_directory: See :py:func:`update_repo`.
+        :param target_bucket: See :py:func:`update_repo`.
+        :param target_key: See :py:func:`update_repo`.
+        :param source_repo: See :py:func:`update_repo`.
+        :param packages: See :py:func:`update_repo`.
+        :param version: See :py:func:`update_repo`.
+        """
+        dispatchers = [aws.get_dispatcher(), yum.get_dispatcher(),
+                       base_dispatcher]
+        sync_perform(
+            ComposedDispatcher(dispatchers),
+            update_repo(rpm_directory, target_bucket, target_key, source_repo,
+                        packages))
+
+    def create_fake_repository(self, files):
+        """
+        Create files in a directory to mimic a repository of packages.
+
+        :param dict source_repo: Dictionary mapping names of files to create to
+            contents.
+        :return: FilePath of directory containing fake package files.
+        """
+        source_repo = FilePath(tempfile.mkdtemp())
+        for key in files:
+            new_file = source_repo.preauthChild(key)
+            if not new_file.parent().exists():
+                new_file.parent().makedirs()
+            new_file.setContent(files[key])
+        return 'file://' + source_repo.path
+
+    def setUp(self):
+        self.scratch_directory = FilePath(tempfile.mkdtemp())
+        self.addCleanup(self.scratch_directory.remove)
+        self.rpm_directory = self.scratch_directory.child(
+            b'distro-version-arch')
+        self.target_key = 'test/target/key'
+        self.target_bucket = 'test-target-bucket'
+        self.build_server = 'http://test-build-server.com'
+        self.packages = ['clusterhq-flocker-cli', 'clusterhq-flocker-node']
```

Comment:
I'll put a note here. I didn't want to include that in the repository because it is significantly bigger than the other binaries. It doesn't seem necessary to have all the right package names here. If you think that it is, I will change this and then use the subset `['clusterhq-flocker-cli', 'clusterhq-flocker-node']` for the test which uses the real yum tools.